lib/writer/index.js

```diff
@@ -0,0 +1,306 @@
+const BlockReadStream = require('./block-read-stream')
+const ChecksumStream = require('./checksum-stream')
+const ProgressStream = require('progress-stream')
+const debug = require('debug')('image-writer')
+const EventEmitter = require('events').EventEmitter
+const _ = require('lodash')
+
+/* eslint-disable no-underscore-dangle */
+
+/**
+ * @summary ImageWriter class
+ * @class
+ */
+class ImageWriter extends EventEmitter {
+  /**
+   * @summary ImageWriter constructor
+   * @param {Object} options - options
+   * @example
+   * new ImageWriter(options)
+   */
+  constructor (options) {
+    super()
+
+    this.options = options
+
+    this.source = null
+    this.pipeline = null
+    this.target = null
+
+    this.hadError = false
+
+    this.bytesRead = 0
+    this.bytesWritten = 0
+    this.checksum = {}
+  }
+
+  /**
+   * @summary Start the writing process
+   * @returns {ImageWriter} imageWriter
+   * @example
+   * imageWriter.write()
+   */
+  write () {
+    this.hadError = false
+
+    this._createWritePipeline(this.options)
+      .on('checksum', (checksum) => {
+        debug('write:checksum', checksum)
+        this.checksum = checksum
+      })
+      .on('error', (error) => {
+        this.hadError = true
+        this.emit('error', error)
+      })
+
+    this.target.on('finish', () => {
+      this.bytesRead = this.source.bytesRead
+      this.bytesWritten = this.target.bytesWritten
+      if (this.options.verify) {
+        this.verify()
+      } else {
+        this._emitFinish()
+      }
+    })
+
+    return this
+  }
+
+  /**
+   * @summary Start the writing process
+   * @returns {ImageWriter} imageWriter
+   * @example
+   * imageWriter.verify()
+   */
+  verify () {
+    this._createVerifyPipeline(this.options)
+      .on('error', (error) => {
+        this.hadError = true
+        this.emit('error', error)
+      })
+      .on('checksum', (checksum) => {
+        debug('verify:checksum', this.checksum, '==', checksum)
+        if (!_.isEqual(this.checksum, checksum)) {
+          this.emit('error', new Error(`Verification failed: ${JSON.stringify(this.checksum)} != ${JSON.stringify(checksum)}`))
+        }
+      })
+      .on('finish', () => {
+        debug('verify:end')
+        this._emitFinish()
+      })
+
+    return this
+  }
+
+  /**
+   * @summary Abort the flashing process
+   * @example
+   * imageWriter.abort()
+   */
+  abort () {
+    if (this.source) {
+      this.emit('abort')
+      this.source.destroy()
+    }
+  }
+
+  /**
+   * @summary Emits the `finish` event with state metadata
+   * @private
+   * @example
+   * this._emitFinish()
+   */
+  _emitFinish () {
+    this.emit('finish', {
+      bytesRead: this.bytesRead,
+      bytesWritten: this.bytesWritten,
+      checksum: this.checksum
+    })
+  }
+
+  /**
+   * @summary Creates a write pipeline from given options
+   * @private
+   * @param {Object} options - options
+   * @param {Object} options.image - source image
```

Comment:
Maybe we can generalize this to `options.source` and `options.target` in the `_.create*Pipeline`s?

Reply:
Yeah, I was planning to do this in the course of using network locations & devices as sources as that'll require a generalization of these anyways. I just kept it as is for now to stay compatible with the wide rest of `image-stream` etc.